isolation-forest/src/main/scala/com/linkedin/relevance/isolationforest/IsolationForest.scala

```diff
@@ -130,24 +130,46 @@ class IsolationForest(override val uid: String) extends Estimator[IsolationFores
     val isolationForestModel = copyValues(
       new IsolationForestModel(uid, isolationTrees, numSamples).setParent(this))
 
+    // Determine and set the model threshold based upon the specified contamination and
+    // contaminationError parameters.
     if ($(contamination) > 0.0) {
       // Score all training instances to determine the threshold required to achieve the desired
       // level of contamination. The approxQuantile method uses the algorithm in this paper:
       // https://dl.acm.org/citation.cfm?id=375670
-      // The relative error was set to ensure the fraction of samples found to be outliers during
-      // training is within 1% of the value of the parameter $(contamination).
-      val outlierScoreThreshold = isolationForestModel
+      val scores = isolationForestModel
         .transform(df)
-        .stat.approxQuantile($(scoreCol), Array(1 - $(contamination)), $(contamination) * 0.01)
+        .map(row => OutlierScore(row.getAs[Double]($(scoreCol))))
+        .cache()
+      val outlierScoreThreshold = scores
+        .stat.approxQuantile("score", Array(1 - $(contamination)), $(contaminationError))
         .head
       isolationForestModel.setOutlierScoreThreshold(outlierScoreThreshold)
+
+      // Determine labels for each instance using the newly calculated threshold and verify that the
+      // fraction of positive labels is in agreement with the user specified contamination.
+      val observedContamination = scores
+        .map(outlierScore => if(outlierScore.score >= outlierScoreThreshold) 1.0 else 0.0)
+        .reduce(_ + _) / scores.count()
+      val effectiveError = if (${contaminationError} == 0.0) {
+        // If the threshold is calculated exactly, then assume a relative 1% error on the specified
+        // contamination for the verification.
+        $(contamination) * 0.01
```

Comment:
This is a good point. These two errors are pretty independent.

The `contaminationError` parameter is actually used in the calculation of the threshold, while the `effectiveError` here is just used for a validation; a warning is issued if the resulting threshold is significantly different than expected.
As you can see in the diff, the old version of the code always specified a `contaminationError` that was 1% of the `contamination` parameter ($(contamination) * 0.01`). This is now a parameter for the user to enter, but in the docs I note this is a good value.

If the user specifies a `contaminationError` then the `effectiveError` used for the validation is the same. If the user selects the default of an "exact" calculation of the threshold (slow and doesn't scale as well), then we assume a 1% error relative to the contamination for the validation calculation. This is a somewhat arbitrary, but seemingly reasonable choice; it just controls under what circumstances a warning will be issued.

In most cases, the user shouldn't have to worry about the `effectiveError` at all (no warnings), so I think mentioning the internal `effectiveError` in the user docs might cause more confusion than otherwise.

I just submitted an update with the following changes:

1) Changed the name `effectiveError` to `verificationError`, which I think is a bit clearer.
2) Added a detailed in-code comment explaining the difference between the errors.
3) Expanded upon the warning message to make it more clear what the user should do.

Reply:
Thanks!!!